### ISSUE TITLE
Creating read_id as column from modified base call (guppy) output

### ISSUE THREAD

**Issue by JayVaz18** [nanoporetech/ont_fast5_api#53]
Hello,

I recently used guppy to find modified base calls and that resulted in a fast5 file. After running this command: 

```
from ont_fast5_api.fast5_interface import get_fast5_file
with get_fast5_file(fast5_filepath, mode="r") as f5:
for read_id in f5.get_read_ids():
read = f5.get_read(read_id)
latest_basecall = read.get_latest_analysis('Basecall_1D')
mod_base_table = read.get_analysis_dataset(
latest_basecall, 'BaseCalled_template/ModBaseProbs')

table_path = '{}/BaseCalled_template/ModBaseProbs'.format(latest_basecall)
metadata = read.get_analysis_attributes(table_path)
print(read_id, metadata['modified_base_long_names'])
print(read_id, mod_base_table)

```

With this, I'm able to run the print(read_id, mod_base_table) to see for each read, the probability of each base being modified. However, I would like to pass this to a Pandas data frame to make it easier to read and would like to have a column for read_id, and then columns for A, 6mA, 5mC, G, T. This is simple to do for mod_base_table but I'm struggling with how to do this for read_id.

Bottom line is, is there a correct way to do this with your package?

**Comment by cjw85** [comment]
Hi @JayVaz18,

You might want to look at [this code](https://github.com/nanoporetech/fast5mod/blob/60b16f7da986c3629ee5605c506635a15a0d5078/fast5mod/programs.py#L114) from the [fast5mod](GitHub.com/nanoporetech/fast5mod) package which shows how to extract the data in a meaningful manner.

**Comment by fbrennen** [comment]
Hi @JayVaz18 -- so you're after something like this?

```
0 <read_id> <A> <6mA> <C> <5mC> <G> <T>
1 <read_id> <A> <6mA> <C> <5mC> <G> <T>
2 <read_id> <A> <6mA> <C> <5mC> <G> <T>
[...]
```
@cjw85 linked you some code to generate the full list of base names -- all you need to do then is finish the job:
```
import pandas as pd

df = pd.DataFrame(mod_base_table, columns=list_of_all_base_names)
df['read_id'] = read_id
```

**Comment by JayVaz18** [comment]
> Hi @JayVaz18,
> 
> You might want to look at [this code](https://github.com/nanoporetech/fast5mod/blob/60b16f7da986c3629ee5605c506635a15a0d5078/fast5mod/programs.py#L114) from the [fast5mod](GitHub.com/nanoporetech/fast5mod) package which shows how to extract the data in a meaningful manner.

Thanks for your input. However, I can't install fast5mod. When I try to use Conda and pip as provided on the fast5mod GitHub, I keep getting the message that fast5mod does't exist.

**Comment by cjw85** [comment]
@JayVaz18,

Are you trying to use Python 3.9? There are packages available on PyPI for Python 3.5-3.8 inclusive. (The 3.7 and 3.8 ones have been added just today!)

**Comment by JayVaz18** [comment]
> @JayVaz18,
> 
> Are you trying to use Python 3.9? There are packages available on PyPI for Python 3.5-3.8 inclusive. (The 3.7 and 3.8 ones have been added just today!)

Ah yes, I am using Python 3.9. I will downgrade to 3.8 and try again.

**Comment by cjw85** [comment]
Can you indicate how you install fast5mod and on what system you are running. If you are running on macOS then I have a suspicion the PyPI source distribution is broken (though I would have expected an error on install); while we fix that you can use the code from the Github repository. On Linux the PyPI packages just worked correctly for me.

**Comment by JayVaz18** [comment]
> Can you indicate how you install fast5mod and on what system you are running. If you are running on macOS then I have a suspicion the PyPI source distribution is broken (though I would have expected an error on install); while we fix that you can use the code from the Github repository. On Linux the PyPI packages just worked correctly for me.

I'm using a Mac Mojave. Here is the Warning I get when running "pip install fast5mod":

```
Successfully built fast5mod pysam progressbar33 mappy
Installing collected packages: numpy, cached-property, pycparser, progressbar33, h5py, pysam, ont-fast5-api, mappy, cffi, fast5mod
WARNING: Value for scheme.headers does not match. Please report this to <https://github.com/pypa/pip/issues/9617>
distutils: /Users/jakevazquez18/anaconda3/envs/py36/include/python3.6m/UNKNOWN
sysconfig: /Users/jakevazquez18/anaconda3/envs/py36/include/python3.6m
WARNING: Additional context:
user = False
home = None
root = None
prefix = None
Successfully installed cached-property-1.5.2 cffi-1.14.5 fast5mod-1.0.2 h5py-3.1.0 mappy-2.18 numpy-1.18.5 ont-fast5-api-3.0.0 progressbar33-2.4 pycparser-2.20 pysam-0.16.0.1
(py36) mski1911:~ jakevazquez18$ fast5mod
Traceback (most recent call last):
  File "/Users/jakevazquez18/anaconda3/envs/py36/bin/fast5mod", line 5, in <module>
    from fast5mod.fast5mod import main
  File "/Users/jakevazquez18/anaconda3/envs/py36/lib/python3.6/site-packages/fast5mod/fast5mod.py", line 7, in <module>
    import fast5mod.programs
  File "/Users/jakevazquez18/anaconda3/envs/py36/lib/python3.6/site-packages/fast5mod/programs.py", line 18, in <module>
    import libfast5mod
ModuleNotFoundError: No module named 'libfast5mod'

```

**Comment by JayVaz18** [comment]
> Can you indicate how you install fast5mod and on what system you are running. If you are running on macOS then I have a suspicion the PyPI source distribution is broken (though I would have expected an error on install); while we fix that you can use the code from the Github repository. On Linux the PyPI packages just worked correctly for me.

Even when I take the code from programs.py, and use the import statements, it still says it can't find libfast5mod.

**Comment by cjw85** [comment]
As I said above, I suspect the PyPI source distribution is broken. I just confirmed this to be true; there are a couple of files missing. How you managed to install without error I am unsure, I would expect an error indicating a file `build.py` is not available.

Regardless, I just created a new version (v1.0.3) which I hope will install correctly.

**Comment by JayVaz18** [comment]
> As I said above, I suspect the PyPI source distribution is broken. I just confirmed this to be true; there are a couple of files missing. How you managed to install without error I am unsure, I would expect an error indicating a file `build.py` is not available.
> 
> Regardless, I just created a new version (v1.0.3) which I hope will install correctly.

Got it! Thanks so much for your help (and patience). :)

**Comment by JayVaz18** [comment]
> As I said above, I suspect the PyPI source distribution is broken. I just confirmed this to be true; there are a couple of files missing. How you managed to install without error I am unsure, I would expect an error indicating a file `build.py` is not available.
> 
> Regardless, I just created a new version (v1.0.3) which I hope will install correctly.

Sorry, one more question. I'm decided to try fast5mod in the terminal and I ran this command:

```
fast5mod guppy2sam /Users/jakevazquez18/Desktop/guppy/fast5/mod_out/workspace/FAO46487_2ec6d299_722.fast5 --workers 8 --io_workers 4 --recursive --reference /Users/jakevazquez18/Desktop/Clemence/ref_genome/mask_genome.fa

```
But got this as a result (the program didn't stop, just paused and never picked back up:

``` 
[15:00:58 - ModExtract] NOTE: Mod. base scores are output w.r.t the sequencing direction, not the aligned read orientation.
@HD	VN:1.5	SO:unsorted
@CO	Guppy basecaller mod. base tags are stored w.r.t. the sequencing direction, they should be reversed for reads aligning to the reverse strand.
@SQ	SN:chr1	LN:230218
@SQ	SN:chr2	LN:813184
@SQ	SN:chr3	LN:316620
@SQ	SN:chr4	LN:1531933
@SQ	SN:chr5	LN:576874
@SQ	SN:chr6	LN:270161
@SQ	SN:chr7	LN:1090940
@SQ	SN:chr8	LN:562643
@SQ	SN:chr9	LN:439888
@SQ	SN:chr10	LN:745751
@SQ	SN:chr11	LN:666816
@SQ	SN:chr12	LN:1078177
@SQ	SN:chr13	LN:924431
@SQ	SN:chr14	LN:784333
@SQ	SN:chr15	LN:1091291
@SQ	SN:chr16	LN:948066
[15:00:58 - ModExtract] Loading alignment index.
[15:01:00 - ModExtract] Alignment index loaded.
[15:01:00 - Extractor] Starting worker processes.
[15:01:00 - Extractor] Found 1 files to process.
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/Users/jakevazquez18/anaconda3/envs/py36/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/Users/jakevazquez18/anaconda3/envs/py36/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/jakevazquez18/anaconda3/envs/py36/lib/python3.6/site-packages/fast5mod/programs.py", line 214, in _run
    if self.queue.qsize() < self.max_size:
  File "/Users/jakevazquez18/anaconda3/envs/py36/lib/python3.6/multiprocessing/queues.py", line 117, in qsize
    return self._maxsize - self._sem._semlock._get_value()
NotImplementedError
```

**Comment by cjw85** [comment]
A search of the internet reveals this is a commonly encountered issue on macOS within the Python standard library. There is not too much we can do about this I'm afraid without reimplementing the Queue class from the standard library.

**Comment by cjw85** [comment]
@JayVaz18 

In v1.0.4 of fast5mod I have reimplemented the standard library Queue class in a way that should work on macOS.

**Comment by JayVaz18** [comment]
> @JayVaz18
> 
> In v1.0.4 of fast5mod I have reimplemented the standard library Queue class in a way that should work on macOS.

This is the error message I get now after upgrading to v1.0.4: 

```
Traceback (most recent call last):
  File "/Users/jakevazquez18/anaconda3/envs/py36/bin/fast5mod", line 5, in <module>
    from fast5mod.fast5mod import main
  File "/Users/jakevazquez18/anaconda3/envs/py36/lib/python3.6/site-packages/fast5mod/fast5mod.py", line 7, in <module>
    import fast5mod.programs
  File "/Users/jakevazquez18/anaconda3/envs/py36/lib/python3.6/site-packages/fast5mod/programs.py", line 16, in <module>
    from fast5mod.executor import ProcessPoolExecutor, Queue, ThreadPoolExecutor
  File "/Users/jakevazquez18/anaconda3/envs/py36/lib/python3.6/site-packages/fast5mod/executor.py", line 74, in <module>
    class Queue(multiprocessing.queues.Queue):
AttributeError: module 'multiprocessing' has no attribute 'queues'
```
